src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HPack/IntegerDecoder.cs

```diff
@@ -41,11 +42,29 @@ public bool StartDecode(byte b, int prefixLength)
         /// <returns>If the integer has been fully decoded, true. Otherwise, false -- <see cref="Decode(byte)"/> must be called on subsequent bytes.</returns>
         public bool Decode(byte b)
         {
-            _i = _i + (b & 127) * (1 << _m);
+            if (_m - BitOperations.LeadingZeroCount(b) > 0)
+            {
+                // A bit would be shifted beyond 31 bits.
+                throw new HPackDecodingException(SR.net_http_hpack_bad_integer);
+            }
+
+            _i = _i + ((b & 127) << _m);
             _m = _m + 7;
 
-            if ((b & 128) != 128)
+            if (_i < 0)
```

Comment:
It seems to me there's a case that's missed here.

Suppose a prefix length of 1, and the following encoded values: 0xFF 0xFF 0xFF 0xFF 0xFF 0x0F.

After StartDecode, _m = 0 and _i = 0x7F.
After 4 calls to Decode, _m = 28 and _i = 0x7F + 0x0FFFFFFF (i.e. 28 ones) == 0x1000007E.

Now, on the fifth call to Decode, we pass in 0x0F as the value. LeadingZeroCount = 28 and _m = 28, so we pass that initial test. But 0x0F << 28 == 0xF0000000 and thus the new value for _i will be 0x0000007E after truncation. This is a positive value, so won't be caught with this < 0 test here.